storage-s3-dynamodb/src/main/java/io/delta/storage/BaseExternalLogStore.java

```diff
@@ -144,83 +166,101 @@ public void write(
             Configuration hadoopConf) throws IOException {
         final FileSystem fs = path.getFileSystem(hadoopConf);
         final Path resolvedPath = stripUserInfo(fs.makeQualified(path));
+        try {
+            // Prevent concurrent writers in this JVM from either
+            // a) concurrently overwriting N.json if overwrite=true
+            // b) both checking if N-1.json exists and performing a "recovery" where they both
+            //    copy T(N-1) into N-1.json
+            //
+            // Note that the mutual exclusion on writing into N.json with overwrite=false from
+            // different JVMs (which is the entire point of BaseExternalLogStore) is provided by the
+            // external cache, not by this lock, of course.
+            //
+            // Also note that this lock path (resolvedPath) is for N.json, while the lock path used
+            // below in the recovery `fixDeltaLog` path is for N-1.json. Thus, no deadlock.
+            pathLock.acquire(resolvedPath);
```

Comment:
Could you remind me why we need to lock the write path? Currently, `PathLock` doesn't support nested locking.

Reply:
We want to lock on the write path since, during the write path, we will copy `T(N) --> N.json`.
As soon as we write `T(N)` and commit `E(N, T(N), complete=false)` a reader could come, see the entry E, and then copy `T(N) --> N.json`.
By having the lock on the write path, we prevent readers from duplicating the copy into N.json.

What do you mean by "nesting"? It is per-path? So we can have a lock on N.json (writer), and then see that N-1.json doesn't exist, so try and grab a lock for N-1.json too to perform the recovery?

Reply:
> we prevent readers from duplicating the copy into N.json.

Is it still possible after we lock `fixDeltaLog`?

> What do you mean by "nesting"?

The following pattern doesn't work when using the same path.
```
pathLock.acquire(resolvedPath);
try {
  pathLock.acquire(resolvedPath);
  try {
  } finally {
    pathLock.release(resolvedPath);
  }
} finally {
  pathLock.release(resolvedPath);
}
```
But I see your point: we lock different paths.

A high-level question: can we just lock `BaseExternalLogStore.copyFile`?

Reply:
> Is it still possible after we lock fixDeltaLog?

Yes. Imagine we have two writers A and B. Imagine A grabs the lock, and B does **not** grab the lock. Then even though A has the lock, B can still write into the same file as A.

Thus, every writer must grab the lock!

> Nesting

Ah yes I see what you mean. But yes we lock on different paths